### PR TITLE
sql: add partial statistics predicate column to `system.table_statistics`

### PR DESCRIPTION
**Pull request by faizaanmadhani**
This commit adds a column to system.table_statistics to store
a string representing the predicate for a partial statistics
collection. Previously, only full table statistics could be
stored in `system.table_statistics`, and this commit adds
support to store partial table statistics to this table. It
assists Epic [CRDB-19449](https://cockroachlabs.atlassian.net/browse/CRDB-19449) and is a prerequisite for adding
support for collecting partial statistics.

Epic: [CRDB-19449](https://cockroachlabs.atlassian.net/browse/CRDB-19449)

Release note (sql change): system.table_statistics now
contains a column called partialPredicate to store a
predicate for a partial statistic collection.

### CODE REVIEW

**Comment by postamar** [comment]
RFC link, for the curious: https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20220126_partial_statistics_collection.md

File: pkg/upgrade/upgrades/upgrades.go [hunk 1]
```diff
@@ -169,11 +169,17 @@ var upgrades = []upgrade.Upgrade{
 		NoPrecondition,
 		addTenantNameColumnAndSystemTenantEntry,
 	),
+
```
Comment (postamar) [inline pkg/upgrade/upgrades/upgrades.go:172]: nit: superfluous newline

**Review by postamar — COMMENTED** [review]
Schema LGTM

**Review by rytaft — COMMENTED** [review]
Nice work! Mostly just some nits.

> Reviewed 11 of 11 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-:-NGIra7O2XAGpaaMS6Ch:b-s1f2li)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @faizaanmadhani and @michae2)

___
*[`-- commits` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIr0h8BQIFY9obREza):*
this is technically a user-visible change, so I'd add a `Release note (sql change)`
___
*[`pkg/clusterversion/cockroach_versions.go` line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGImwit13eWV_i9Ok2G:-NGImwit13eWV_i9Ok2H:bfaruzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/clusterversion/cockroach_versions.go#L317)):*
> ```Go
> 	V23_1Start
> 
> 	// TenantNames adds a name column to system.tenants.
> ```

(Not your change, but this should be V23_1TenantNames)
___
*[`pkg/clusterversion/cockroach_versions.go` line 325 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGImv7CAkHrSAm3jYSW:-NGImv7CAkHrSAm3jYSX:b-qr8fhz) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/clusterversion/cockroach_versions.go#L325)):*
> ```Go
> 	V23_1DescIDSequenceForSystemTenant
> 
> 	// AddPartialPredicateCol adds a column to store the predicate
> ```

nit: V23_1AddPartialPredicateCol

Also, I'd call this V23_1AddPartialStatisticsPredicateCol (or V23_1AddTableStatisticsPartialPredicateCol) to make it really clear what this is.
___
*[`pkg/clusterversion/cockroach_versions.go` line 326 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGImm5P2yam4sMtZ_4a:-NGImm5P2yam4sMtZ_4b:b-qorz7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/clusterversion/cockroach_versions.go#L326)):*
> ```Go
> 
> 	// AddPartialPredicateCol adds a column to store the predicate
> 	// for a partial statistics collection
> ```

nit: end comments with a period.
___
*[`pkg/upgrade/upgrades/alter_system_table_partial_predicate.go` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIpPdI5Yr8rZA5UgDW:-NGIpPdJEyUhYWLjPHrR:baw6tqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/upgrade/upgrades/alter_system_table_partial_predicate.go#L1)):*
> ```Go
> // Copyright 2022 The Cockroach Authors.
> ```

I'd name this file `alter_table_statistics_partial_predicate.go` for consistency with some of the other files (doesn't seem like there is one consistent naming scheme, but I think referencing table_statistics is still better)
___
*[`pkg/upgrade/upgrades/alter_system_table_partial_predicate.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIo4MI2Sv5T5YPMYiW:-NGIo4MI2Sv5T5YPMYiX:b-bmf674) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/upgrade/upgrades/alter_system_table_partial_predicate.go#L33)):*
> ```Go
> 	op := operation{
> 		name:           "add-table-statistics-partialPredicate-col",
> 		schemaList:     []string{"total_consumption"},
> ```

total_consumption -> partialPredicate
___
*[`pkg/upgrade/upgrades/alter_system_table_partial_predicate_test.go` line 64 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIpwKS6Y3XXkptB5Er:-NGIpwKS6Y3XXkptB5Es:bz2pejk) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/upgrade/upgrades/alter_system_table_partial_predicate_test.go#L64)):*
> ```Go
> 	)
> 
> 	// Inject the old copy of the descriptor
> ```

nit: end comment with period (here and below).
___
*[`pkg/upgrade/upgrades/alter_system_table_partial_predicate_test.go` line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIq_b17EjTBZqeQXYS:-NGIq_b17EjTBZqeQXYT:b9r5scl) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/upgrade/upgrades/alter_system_table_partial_predicate_test.go#L146)):*
> ```Go
> 		NextFamilyID: 1,
> 		PrimaryIndex: descpb.IndexDescriptor{
> 			Name:                "primary",
> ```

nit: can use tabledesc.LegacyPrimaryKeyIndexName

**Comment by rytaft** [comment]
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIrqGs3Ir9KwHe6L14):*
nit: table_statistics

**Review by faizaanmadhani — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91248)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rytaft)

___
*[`-- commits` line 14 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIr0h8BQIFY9obREza):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

this is technically a user-visible change, so I'd add a `Release note (sql change)`
</blockquote></details>

Done.
___
*[`pkg/clusterversion/cockroach_versions.go` line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGImwit13eWV_i9Ok2G:-NGJ0Qn-5dGYXrrtiRML:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/clusterversion/cockroach_versions.go#L317)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

(Not your change, but this should be V23_1TenantNames)
</blockquote></details>

Done.
___
*[`pkg/clusterversion/cockroach_versions.go` line 325 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGImv7CAkHrSAm3jYSW:-NGJ2KUt9lElfObnOJcf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/clusterversion/cockroach_versions.go#L325)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: V23_1AddPartialPredicateCol

Also, I'd call this V23_1AddPartialStatisticsPredicateCol (or V23_1AddTableStatisticsPartialPredicateCol) to make it really clear what this is.
</blockquote></details>

Done.
___
*[`pkg/clusterversion/cockroach_versions.go` line 326 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGImm5P2yam4sMtZ_4a:-NGJ2WgJ7P2XNj-2PXzs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/clusterversion/cockroach_versions.go#L326)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: end comments with a period.
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/alter_system_table_partial_predicate.go` line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIpPdI5Yr8rZA5UgDW:-NGJ2DY-4dKD4iiYF0Cb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/upgrade/upgrades/alter_system_table_partial_predicate.go#L1)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I'd name this file `alter_table_statistics_partial_predicate.go` for consistency with some of the other files (doesn't seem like there is one consistent naming scheme, but I think referencing table_statistics is still better)
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/alter_system_table_partial_predicate.go` line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIo4MI2Sv5T5YPMYiW:-NGJ2CIJ5bSP6R4MsZAS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/upgrade/upgrades/alter_system_table_partial_predicate.go#L33)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

total_consumption -> partialPredicate
</blockquote></details>

Done.
___
*[`pkg/upgrade/upgrades/alter_system_table_partial_predicate_test.go` line 146 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGIq_b17EjTBZqeQXYS:-NGJ2g5V5X2IIGYosUZD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/117fa59dd3c2b502ed6cff4874de09cbb4d3e110/pkg/upgrade/upgrades/alter_system_table_partial_predicate_test.go#L146)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: can use tabledesc.LegacyPrimaryKeyIndexName
</blockquote></details>

Done.

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 8 of 11 files at r3, 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-:-NGJ5SXtEDNSylssgPGG:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2)

**Review by michae2 — APPROVED** [review]
:lgtm: Nice!

> Reviewed 5 of 11 files at r1, 4 of 11 files at r3, 2 of 3 files at r4, 1 of 1 files at r5, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-:-NGJne-i0i3jc9p_xL40:b-8gk8uj)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @faizaanmadhani)

___
*[`pkg/ccl/logictestccl/tests/fakedist/BUILD.bazel` line 13 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/91248#-NGJmfVvBDQSqaYGWsx-:-NGJmfVvBDQSqaYGWsx0:btyfur3) ([raw file](https://github.com/cockroachdb/cockroach/blob/eaaf2a97bf43c9be6964b575684c340cc287a056/pkg/ccl/logictestccl/tests/fakedist/BUILD.bazel#L13)):*
> ```Starlark
>         "//pkg/ccl/logictestccl:testdata",  # keep
>     ],
>     shard_count = 7,
> ```

Did these come from `./dev gen bazel --short`? I'm not sure why they changed.

**Comment by faizaanmadhani** [comment]
> ./dev gen bazel --short

Yes, I'm not too sure what caused them to change, but I'm gonna try rebasing with master and rerunning the build.

**Comment by faizaanmadhani** [comment]
TFTRs! 🎉 

bors r=rytaft,michae2

**Pull request merged**